### ISSUE TITLE
table_exists? returns incorrect results on MySQL

### ISSUE THREAD

**Issue by al2o3cr** [rails/rails#17897]
`table_exists?` can return incorrect results on MySQL if you have tables with `LIKE` metacharaters in their names.

I'm attempting to load a legacy schema, and there's a number of odd naming conventions going on. Most table names end in an underscore (for *reasons*), which causes `table_exists?` to sometimes return true when it shouldn't - breaking things like `force: true` in migrations.

For concreteness, imagine we have a table called `Foos` and another called `foo_` (not their real names :) ). The `create_table` for Foos executes without incident, but the one for `foo_` blows up with 'Unknown table 'foo_': DROP TABLE 'foos_'`.

Under the hood, calling `ActiveRecord::Base.connection.table_exists?('foo_')` runs the following SQL:
```
SHOW TABLES LIKE 'foo_'
```
which, thanks to MySQL's case-insensitive matching and `_` meaning "match any single character`, successfully finds the `Foos` table.

The `LIKE` behavior was introduced in https://github.com/rails/rails/commit/38703ac8972c7e8f3f3f1ac95aa506cc4ae30ef0 back in 2011. My guess is that `LIKE` was used because `SHOW TABLES` doesn't have a particularly clean way to select an exact-match table. @jonleighton may remember more.

I can think of two possibilities for fixing this:

* escape `LIKE` metacharacters in the `tables` function. No idea if anybody's actually expecting to be able to pass them, though.

* do a more-detailed query against `information_schema.tables`. This only works on MySQL 5.0 or higher, but I believe the API adapter gems also enforce this minimum version.

I have a patch, but need to get it working with the existing AR tests etc.

**Comment by kaihuang201** [comment]
I successfully reproduced the behavior with this script https://gist.github.com/5793b720de92cdc4fcd5.git

**Comment by al2o3cr** [comment]
@kaihuang201 Escaping seems like the wrong approach. I've got a patch that replaces the `SHOW TABLES LIKE` query with one against `information_schema` with plain `=`. Will load it up once I get off this eternal airline customer service phone call...

**Comment by kaihuang201** [comment]
@al2o3cr Could you elaborate a little bit on "...one against information_schema with plain =."?

**Comment by al2o3cr** [comment]
I'm preparing a PR for this, but ran into some incidental test breakage. Here's the code:
```ruby
      def tables(name = nil, database = nil, like = nil) #:nodoc:
        database ||= current_database

        sql = "SELECT table_name FROM information_schema.tables "
        sql << "WHERE table_schema = #{quote(database)}"
        sql << " AND table_name = #{quote(like)}" if like

        execute_and_free(sql, 'SCHEMA') do |result|
          result.collect(&:first)
        end
      end
```
and the test:
```ruby
      def test_table_exists_with_like_metacharacters
        assert_not @conn.table_exists?('aut_ors'), 'nonexistent table should not exist'
      end
```
(this goes in `test/cases/adapters/mysql/mysql_adapter_test.rb`)

Tests that are broken by this:

* `ActiveRecord::ConnectionAdapters::MysqlAdapterTest#test_tables_quoting` - broken because the new method returns `[]` instead of raising when asked about a nonexistent database. That test was added in the fix for #5270, but it should probably just be rewritten.

* `BulkAlterTableMigrationsTest#test_adding_indexes` (and the corresponding test for removing indexes) - broken because `assert_queries` counts the SQL generated by `current_database` and the query against `information_schema`. Need to update the behavior of `SQLCounter` for MySQL to ignore the queries generated by `table_exists?`.

Should be able to wrap this up this weekend.
